### ISSUE TITLE
question for INDEL variant calling

### ISSUE THREAD

**Issue by sophienguyen01** [google/deepvariant#813]
I want to use DeepVariant to get INDEL variants only, is there a way to achieve that with `run_deepvariant` command 
 or `make_example` command?

Thank you

**Comment by danielecook** [comment]
When running `make_examples`, you can pass the the following flag: `--select_variant_types='indels'`

**Comment by sophienguyen01** [comment]
thank you @danielecook ,

How is `--select_variant_types='indels'` different from `types_to_alt_align` parameter? 

I want to train DeepVariant on INDEL variants with particular length, is there a command in `make_examples` to generate labeled example for these particular variants? 

Thank you

**Comment by danielecook** [comment]
`types_to_alt_align` refers to the type of variants in which we perform alignments against the alternative variant, when you have also set the `alt_aligned_pileup` flag. 

You might be able to accomplish something like this by making use of the vcf candidate importer. See `--truth_variants` + `--variant_caller=vcf_candidate_importer`

I would expect that if you perform filtering similarly on your training and test data, that this could be a way to develop a model specific to certain size INDEL variants, but we have never tried to do something like this.

**Comment by sophienguyen01** [comment]
Hi @danielecook , I run the `run_deepvariant` command with and without `--select_variant_types='indels'` parameter and observed different number of INDELS between two outputs. My understanding is that this command will remove all the SNP candidates but surprisingly, it also lowers the number of INDELS variants. I attached the visual report of deepvariant result with and without this parameter.

When all type of variants are considered:
![HG003_all visual_report](https://github.com/google/deepvariant/assets/135994439/d8281b58-1d81-4cee-bf76-8e06288d702f)

Only INDELS variants are reported:
![HG003_indels visual_report](https://github.com/google/deepvariant/assets/135994439/ebbe5a77-2a60-4bbc-8aeb-dd2dbc9381ca)

**Comment by kishwarshafin** [comment]
Hi @sophienguyen01,

Can you please share the VCF files so I can look at it? I think the issue would be multi-allelic sites, but I want to confirm that.

**Comment by sophienguyen01** [comment]
The files are big (124MB) in size so I cannot attach here. Is there an email I can send to?

**Comment by kishwarshafin** [comment]
yes, please send it to shafin@google.com

**Comment by kishwarshafin** [comment]
@sophienguyen01 ,

I think I was able to get to the problem but unfortunately, I am unable to reproduce the issue. 

Here's what I did:

Extract indels from the full file:

```bash
bcftools view -v indels HG003_043024.vcf.gz > HG003_043024.indels_only.bcftools_filter.vcf
```

Run stats:
```
bcftools stats HG003_043024.indels_only.bcftools_filter.vcf.gz | grep 'indels:'
SN	0	number of indels:	1240956
```

Compared to parameter-based:
```bash
SN	0	number of indels:	1056774
```

So, we are looking for 184182 variants.

So do subtract:
```bash
bedtools subtract \
-a HG003_043024.indels_only.bcftools_filter.vcf.gz \
-b HG003_indels_043024.vcf.gz | wc -l

180869
```

So roughly it matches.

Now look at some variants:
```bash
bedtools subtract \
-a HG003_043024.indels_only.bcftools_filter.vcf.gz \
-b HG003_indels_043024.vcf.gz | head

chr1	10247	.	TAAACCCTA	T	0.5	RefCall	.	GT:GQ:DP:AD:VAF:PL	./.:9:41:28,4:0.097561:0,14,10
chr1	98999	.	TTTTATTTA	T,TTTTATTTATTTA	20	PASS	.	GT:GQ:DP:AD:VAF:PL	1/2:10:31:20,9,2:0.290323,0.0645161:19,16,12,16,0,21
chr1	99092	.	C	CT	2.7	RefCall	.	GT:GQ:DP:AD:VAF:PL	./.:3:50:19,7:0.14:0,1,8
chr1	101674	.	C	CAAA	0.6	RefCall	.	GT:GQ:DP:AD:VAF:PL	./.:9:29:23,2:0.0689655:0,8,17
chr1	104160	.	A	AACAC,AACACACAC	15.1	PASS	.	GT:GQ:DP:AD:VAF:PL	1/2:5:79:1,37,21:0.468354,0.265823:13,14,6,14,0,9
chr1	108545	.	C	CA	2.7	RefCall	.	GT:GQ:DP:AD:VAF:PL	./.:3:44:12,21:0.477273:0,1,6
chr1	109575	.	CGT	C,CGTGTGT	13	PASS	.	GT:GQ:DP:AD:VAF:PL	1/2:4:22:0,8,10:0.363636,0.454545:11,13,15,13,0,4
chr1	111513	.	C	CTA	19.3	PASS	.	GT:GQ:DP:AD:VAF:PL	1/1:18:33:0,30:0.909091:19,22,0
chr1	180150	.	AC	A,GC	15	PASS	.	GT:GQ:DP:AD:VAF:PL	1/2:2:19:2,6,9:0.315789,0.473684:11,13,2,13,0,1
chr1	180174	.	TAA	T	3.5	PASS	.	GT:GQ:DP:AD:VAF:PL	1/1:3:14:7,4:0.285714:0,9,0
```

So there are few variants that we are not picking up.



Next, I picked the region where variant "chr1 10247" is and ran make_examples with a debug command:

Without filter command:
```bash
chr1 10240 T ['TA']
chr1 10246 TA ['T']
chr1 10249 A ['C']
chr1 10253 TA ['T']
chr1 10256 A ['C']
```
I see these five variants.

With filtering I see:
```bash
FILTERING CANDIDATES
chr1 10240 T ['TA']
chr1 10246 TA ['T']
chr1 10253 TA ['T']
```

I am unsure how to reproduce this. Are you using a publicly available bam file? I can also run DV with and without this command and generate results to investigate further. It would be faster and helpful if you can point me to the bam you are using so it's more specific to your issue.

**Comment by sophienguyen01** [comment]
Hi @kishwarshafin,

It's a HG003 sample, I believe you also use this sample for training DeepVariant. I used our internal HG003 cram file.

If you can send me a link to a public HG003 bam/cram file, I can rerun DV and see if the issue still persists.

Thank you!

**Comment by kishwarshafin** [comment]
@sophienguyen01 , I can run a chr20 test on our end try to replicate.

Looking at the code, the filtering logics are implemented [here](https://github.com/google/deepvariant/blob/r1.6.1/deepvariant/make_examples_core.py#L123).

```python
def _select_biallelic_snps(v):
  return variant_utils.is_snp(v) and variant_utils.is_biallelic(v)


def _select_biallelic_indels(v):
  return variant_utils.is_indel(v) and variant_utils.is_biallelic(v)


def _select_biallelic_insertions(v):
  return variant_utils.has_insertion(v) and variant_utils.is_biallelic(v)


def _select_biallelic_deletions(v):
  return variant_utils.has_deletion(v) and variant_utils.is_biallelic(v)


VARIANT_TYPE_SELECTORS = {
    'snps': _select_biallelic_snps,
    'indels': _select_biallelic_indels,
    'insertions': _select_biallelic_insertions,
    'deletions': _select_biallelic_deletions,
    'multi-allelics': variant_utils.is_multiallelic,
    'all': lambda v: True,
}
```

And the filtering logic is implemented [here](https://github.com/google/deepvariant/blob/r1.6.1/deepvariant/make_examples_core.py#L905). 

It looks like `--select_variant_types='indels multi-allelics'` will give you all multi-allelic indels too. I am unsure if it will solve the issue because the VCF you provided before also misses bi-allelic indels. I will need to debug it further to see if there's something missing. Meanwhile, you can use `--select_variant_types='indels multi-allelics'` to see if it fixes your issue.

**Comment by sophienguyen01** [comment]
Hi,

I tried `--select_variant_types='indels multi-allelics'`, the vcf includes all the INDELS from vcf output that is not included with `--select_variant_types`. However, the output vcf using  `--select_variant_types='indels multi-allelics'` also includes SNPs variants. 

My purpose is to create INDELs training examples only and `--select_variant_types='indels multi-allelics'`  still contains SNPs examples.  Is there a way to filter examples (from tfrecord???.gz files) after `make_examples` step but before `shuffling` step.

I also tried `--truth_variants` (using the truth vcf that only contains INDEL variants) and `-variant_caller=vcf_candidate_importer` parameter, but the number of examples are reduced significantly.
